python2-pybind11/challenge/containers.cpp

```diff
@@ -0,0 +1,62 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "basics.hpp"
+#include "containers.hpp"
+
+namespace py = pybind11;
+
+PYBIND11_DECLARE_HOLDER_TYPE(T, std::shared_ptr<T>);
+
+namespace containers {
+
+class DoodadSetIterator {
+public:
+    /* Keep copy of DoodadSet elements (as shared_ptrs)
+     * This is needed because the iterator might outlive the container.
+     * A more elegant solution would be to have DoodadSet inherit from
+     * std::enable_shared_from_this<DoodadSet>, so that an
+     * iterator can keep the container alive while it is still in use
+     * even though the container has been garbage collected on the Python
+     * side, but that would require modifying the C++ implementation. */
```

Comment:
Ok, so it's not *quite* what we need, but it's very close.  `reference_internal` does two things:
 - it says that the new object (the iterator in our case) is not "owned", so it should not be deleted by pybind11 when its Python reference count goes to zero;
 - it increments the Python reference count of the parent (the container in our case), and will decrement it later.
We only want the latter - because the iterator's own destructor does need to get called by pybind11.

And, reading your first reply on this again, I see that you had already figured this out and I'd just misunderstood what you were saying.  I do think it'd be worthwhile asking for a new return value policy to handle this - not as anything that would delay this ticket or the writeup, but it's clearly something we think pybind11 should have.